src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,179 @@
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            add(value);
+        } else {
+            Node<T> newNode = new Node<>(null, value, null);
+            Node<T> curr = getNode(index);
+            Node<T> prev = curr.prev;
+            prev.next = newNode;
+            newNode.prev = prev;
+            newNode.next = curr;
+            curr.prev = newNode;
+            size++;
+        }
+    }
+
+    private void addFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> node = null;
+        if (index < size / 2) {
+            node = head;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+        } else {
+            node = tail;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+        }
+        return node;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> node = getNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> node = getNode(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> node = getNode(index);
+        final T value = node.value;
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+        if (node.next == null) {
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+        size--;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> nodeToRemove = new Node<>(null, object, null);
+        int temp = 0;
+        for (Node<T> i = head; i != null; i = i.next) {
+            temp++;
+            if (nodeToRemove.value == i.value
+                    || nodeToRemove.value != null && nodeToRemove.value
+                    .equals(i.value)) {
```

Comment:
it doesn't seem necessary to split line for .equals() call, it only makes reading harder